### ISSUE TITLE
Move istioctl to its own repository

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#17398]
Rationale:
* istioctl is logically stand alone
* istioctl depends on operator. Operator depends on k8s.io/kubernetes. Depending on kubernetes is explicitly frowned upon, so much the kubernetes ships a broken go.mod file to stop people (but we still find a way...).
* Operator also depends on helm, which depends on k8s.io/kubernetes 1.15. Helm will not update to 1.16 for the forseeable future (I reached out to helm team to confirm), meaning Istio cannot support k8s 1.16 features, meaning we cannot support 1.16 clusters that use EndpointSlice (https://github.com/istio/istio/issues/17158)

Challenges:
* We need to update our release process to handle this
* istioctl depends on pilot directly -- this means for the time being istioctl is actually above istio/istio in the dependency graph. Ideally it is either below it or completely isolated from this

**Comment by ayj** [comment]
+1 to the general idea. 

What's the timeline for adopting 1.16/EndpointSlice?

**Comment by kyessenov** [comment]
Last time I looked into this, `istioctl validate` was bringing in a ton of code from mixer and pilot.

**Comment by yangminzhu** [comment]
> Last time I looked into this, `istioctl validate` was bringing in a ton of code from mixer and pilot.

I think some other commands may also be using internal data structure and code from pilot and galley.

**Comment by howardjohn** [comment]
yes istioctl depends on a lot of other code for all sorts of things. That is why I suggested that istioctl would need to import istio/istio. 

The timeline is hopefully in 1.4, because in k8s 1.16 this feature is added, and it replaced Endpoint rather than extending it - so if they turn it on istio breaks. But it's alpha in 1.16 so it's probably ok to do by 1.5

**Comment by ayj** [comment]
@ozevren, do the analyzers (in)directly depend on any k8s code, or is completely based on collections and our own protos?

How much work is it to update the release process? We can fix the dependency inversion post 1.4 if necessary.

**Comment by ayj** [comment]
Will istioctl inherit the k8s 1.16 dependency from istio/istio when it imports?

**Comment by howardjohn** [comment]
> Will istioctl inherit the k8s 1.16 dependency from istio/istio when it imports?

Ahh yeah that is a great point.. I think that will just invert the problem so now istioctl is broken instead of istio/istio :confused:. So this may not actually be feasible.

If I have some time I'll give it a shot of just doing it locally and seeing what happens maybe

**Comment by ozevren** [comment]
The core/majority of the analyzer code does not depend on K8s code, however it depends on K8s libraries to read and parse K8s based resources.

**Comment by ayj** [comment]
istio.io/istio/pkg/kube/inject/webhook_test.go also depends on helm.

**Comment by howardjohn** [comment]
> istio.io/istio/pkg/kube/inject/webhook_test.go also depends on helm.

It does but must depend on a smaller subset of the repo. as a POC I delete `istioctl/` and see our dependency on k8s.io/kubernetes is dropped

**Comment by ayj** [comment]
I look at our dependencies on the k8s.io/kubernetes repo. It should be straightforward to remove through a combination of k8s.io/kubectl/pkg and refactoring. Does that solve the problem or do we still need to eliminate helm as well?

**Comment by howardjohn** [comment]
Not super familiar with go dependencies. I tried removing the direct usages in operator repo and there was still the indirect import, which I think was from helm. But it does appear we can depend on helm without pulling that in, as we did in the webhook test. So it is possible? Although I am not sure how many changes it would require in operator to drop the dependency fully.

`go mod graph` is a useful command I learned after I tried this, so may have missed some useful information

**Comment by howardjohn** [comment]
Just another data point, I wanted to start trying out stuff that needs 1.16, so I deleted `istioctl/` folder. There were a few things that depend on istioctl but nothing major -- as a short term hack I just deleted these, but it would be feasible to drop the dependencies on istioctl. After I dropped istioctl I was able to successfully upgrade to 1.16.

That being said, if we can just trim some dependencies in operator it may be a lot less painful for now

**Comment by howardjohn** [comment]
Forgot I created a similar issue for this: https://github.com/istio/istio/issues/17245

**Comment by pitlv2109** [comment]
+1 for the change. I think it'll help driving community contribution as well, given that the istioctl codebase is not as scary and complex as istio.

**Comment by duderino** [comment]
@sushicw not sure if you want to take this one, but might make it easier for people to use istioctl analyze

**Comment by esnible** [comment]
I am not sure how to triage this.  P1?  P2?  P3?

**Comment by howardjohn** [comment]
This is P2 at best I think. Its arguable if we want mono-repo or multi-repo due to the complexity it brings

**Comment by howardjohn** [comment]
given current plans this seems like we won't do this
